### PR TITLE
Show private channels in autocomplete

### PR DESCRIPTION
**Pull request by agnivade**
This is supported in all Engines:
MySQL, Postgres, Bleve, Elasticsearch.

https://mattermost.atlassian.net/browse/MM-18496

```release-note
Private channels will now appear in channel autocomplete.
If you are using Bleve or ElasticSearch, you will have to reindex
the channels again to populate them with the new attributes.
```

A large chunk of this work has been based on the earlier
effort at #17804.

Full credit goes to @arvinDarmawan.

### CODE REVIEW

File: services/searchengine/bleveengine/search.go [hunk 1]
```diff
@@ -303,21 +303,39 @@ func (b *BleveEngine) DeletePost(post *model.Post) *model.AppError {
...
+	userIDQ := bleve.NewTermQuery(userID)
+	userIDQ.SetField("UserIds")
+	userQ.AddMust(userIDQ)
+	userQ.AddMust(privateQ)
+
+	channelDisjunctionQ := bleve.NewDisjunctionQuery()
+	channelDisjunctionQ.AddQuery(teamQ) // teamId && != 'P'
+	channelDisjunctionQ.AddQuery(userQ) // userID && 'p'
+
+	queries := []query.Query{channelDisjunctionQ}
```
Comment (agnivade) [inline services/searchengine/bleveengine/search.go:338]: @mgdelacroix - help needed on this part. The logic that I want is

```
teamID = <>
AND
term prefixMatch <>
AND
(type != P OR (type == P AND <> in userids))
```

Comment (agnivade) [reply services/searchengine/bleveengine/search.go:338]: I was able to improve to this:

```
teamIdQ := bleve.NewTermQuery(teamId)
	teamIdQ.SetField("TeamId")
	queries := []query.Query{teamIdQ}

	boolNotPrivate := bleve.NewBooleanQuery()
	privateQ := bleve.NewTermQuery(string(model.ChannelTypePrivate))
	privateQ.SetField("Type")
	boolNotPrivate.AddMustNot(privateQ)

	userQ := bleve.NewBooleanQuery()
	userIDQ := bleve.NewTermQuery(userID)
	userIDQ.SetField("UserIDs")
	userQ.AddMust(userIDQ)
	userQ.AddMust(privateQ)

	channelTypeQ := bleve.NewDisjunctionQuery()
	channelTypeQ.AddQuery(boolNotPrivate)
	channelTypeQ.AddQuery(userQ) // userID && 'p'

	queries = append(queries, channelTypeQ)

	if term != "" {
		nameSuggestQ := bleve.NewPrefixQuery(strings.ToLower(term))
		nameSuggestQ.SetField("NameSuggest")
		queries = append(queries, nameSuggestQ)
	}
```

But it still shows only public channels.

Comment (agnivade) [reply services/searchengine/bleveengine/search.go:338]: Nvm, I figured it out.

**Comment by agnivade** [comment]
I have another PR following from this which extends it to search channels across teams. Feel free to leave comments in the PR, but I will address them in the final PR and close this one for simplicity's sake.

**Review by mgdelacroix — APPROVED** [review]
Awesome job!! added a couple comments, mostly nits, nothing blocking

**Comment by agnivade** [comment]
Thanks @mgdelacroix! I will close this one and incorporate all the feedback in the next PR which adds cross team channel search, which is basically a continuation of this work.

**Pull request closed without merging**